src/par_iter/internal.rs

```diff
@@ -28,6 +28,20 @@ pub trait Producer: IntoIterator + Send + Sized {
     /// Split into two producers; one produces items `0..index`, the
     /// other `index..N`. Index must be less than `N`.
     fn split_at(self, index: usize) -> (Self, Self);
+
+    /// Iterate the whole producer, feeding each element into `folder`,
+    /// stopping when it is full.
+    fn fold_using<F>(self, mut folder: F) -> F::Result
+        where F: Folder<Self::Item>,
+    {
+        for item in self {
+            folder = folder.consume(item);
+            if folder.full() {
+                break;
+            }
+        }
+        folder.complete()
```

Comment:
I think it may be more useful to leave `complete` out of this.  At least for `ChainProducer`, that would let it call `fold_using` separately on its two parts, recursively if there are multiple levels of chains, then just let the bridge functions complete it at the end.